block-producer/src/block_builder/kernel.rs

```diff
@@ -0,0 +1,239 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use miden_air::ExecutionOptions;
+use miden_node_proto::domain::{AccountInputRecord, BlockInputs};
+use miden_objects::{
+    accounts::AccountId, assembly::Assembler, crypto::merkle::MerkleStore, BlockHeader, Digest,
+    Felt,
+};
+use miden_stdlib::StdLibrary;
+use miden_vm::{
+    crypto::MerkleError, execute, AdviceInputs, DefaultHost, ExecutionError, MemAdviceProvider,
+    Program, StackInputs,
+};
+use thiserror::Error;
+
+use crate::SharedTxBatch;
+
+use super::BuildBlockError;
+
+#[derive(Error, Debug)]
+pub enum BlockKernelError {
+    #[error("Received invalid merkle path")]
+    InvalidMerklePaths(MerkleError),
+    #[error("program execution failed")]
+    ProgramExecutionFailed(ExecutionError),
+    #[error("invalid return value on stack (not a hash)")]
+    InvalidRootReturned,
+}
+
+/// Note: For now, the "block kernel" only computes the account root. Eventually, it will compute
+/// the entire block header.
+///
+/// Stack inputs: [num_accounts_updated, OLD_ACCOUNT_ROOT, NEW_ACCOUNT_HASH_0, account_id_0, ... ,
+/// NEW_ACCOUNT_HASH_n, account_id_n]
+const BLOCK_KERNEL_MASM: &str = "
+use.std::collections::smt64
+
+begin
+    push.1
+    while.true
+        # stack: [counter, ROOT_0, ..., NEW_ACCOUNT_HASH_i, account_id_i , ...]
+
+        # Move counter down for next iteration
+        movdn.9
+        # => [ROOT_i, NEW_ACCOUNT_HASH_i, account_id_i, counter, ...]
+
+        # Move root down (for smt64.set)
+        movdn.8 movdn.8 movdn.8 movdn.8
+        # => [NEW_ACCOUNT_HASH_i, account_id_i, ROOT_i, counter, ...]
+
+        # set new value in SMT
+        smt64.set dropw
+        # => [ROOT_{i+1}, counter, ...]
+
+        # loop counter
+        movup.4 sub.1 dup neq.0
+        # => [0 or 1, counter-1, ROOT_{i+1}, ...]
+    end
+
+    drop
+    # => [ROOT_{n-1}]
+end
+";
+
+#[derive(Debug)]
+pub(super) struct BlockKernel {
+    program: Program,
+}
+
+impl BlockKernel {
+    pub fn new() -> Self {
+        let account_program = {
+            let assembler = Assembler::default()
+                .with_library(&StdLibrary::default())
+                .expect("failed to load std-lib");
+
+            assembler
+                .compile(BLOCK_KERNEL_MASM)
+                .expect("failed to load account update program")
+        };
+
+        Self {
+            program: account_program,
+        }
+    }
+
+    // Note: this will eventually all be done in the VM
+    pub fn compute_block_header(
+        &self,
+        witness: BlockHeaderWitness,
+    ) -> Result<BlockHeader, BuildBlockError> {
```

Comment:
I think this is fine for now, but eventually this method will probably look something like this:
```Rust
pub fn prove(
    &self,
    witness: BlockWitness,
) -> Result<(BlockHeader, ExecutionProof), BuildBlockError> 
```
Basically, we'd compute the block header and also prove that it was computed correctly.

Reply:
Changed the name to `prove` and adjusted the comment